src/model/backup.ts

```diff
@@ -340,50 +333,33 @@ export async function addWalletToCloudBackup({
 }
 
 // we decrypt seedphrase and private key before backing up
-export async function decryptAllPinEncryptedSecretsIfNeeded(secrets: Record<string, string>, userPIN?: string) {
+export async function decryptAllPinEncryptedSecretsIfNeeded(secrets: Record<string, string>, maybeUserPIN?: string) {
   const processedSecrets = { ...secrets };
   // We need to decrypt PIN code encrypted secrets before backup
-  const hasBiometricsEnabled = await kc.getSupportedBiometryType();
-  if (IS_ANDROID && !hasBiometricsEnabled) {
-    /*
-     * The PIN code is passed as an argument.
-     * Authentication is handled at the call site.
-     * If we don't have PIN information, we throw an error.
-     * Both for the developer and the user if something goes wrong.
-     */
-    if (userPIN === undefined) {
-      throw new Error(CLOUD_BACKUP_ERRORS.MISSING_PIN);
+  if (IS_ANDROID) {
+    let userPIN = maybeUserPIN;
+    // We only prompt for PIN if it is currently needed, but it is possible
+    // that secrets were previously encrypted with PIN, so we also need
+    // to prompt for PIN here if needed.
+    if (!userPIN && Object.values(processedSecrets).some(secret => secret.includes('cipher'))) {
+      try {
+        // eslint-disable-next-line require-atomic-updates
+        userPIN = await authenticateWithPIN();
+      } catch (e) {
+        throw new Error(CLOUD_BACKUP_ERRORS.WRONG_PIN);
+      }
     }
-
     // We go through each secret here and try to decrypt it if it's needed
     await Promise.all(
       Object.keys(processedSecrets).map(async key => {
         const secret = processedSecrets[key];
         const theKeyIsASeedPhrase = endsWith(key, seedPhraseKey);
         const theKeyIsAPrivateKey = endsWith(key, privateKeyKey);
 
-        if (theKeyIsASeedPhrase) {
-          const parsedSecret = JSON.parse(secret);
-          const seedphrase = parsedSecret.seedphrase;
-
-          if (userPIN && seedphrase && seedphrase?.includes('cipher')) {
-            const decryptedSeedPhrase = await encryptor.decrypt(userPIN, seedphrase);
-            processedSecrets[key] = JSON.stringify({
-              ...parsedSecret,
-              seedphrase: decryptedSeedPhrase,
-            });
-          }
-        } else if (theKeyIsAPrivateKey) {
-          const parsedSecret = JSON.parse(secret);
-          const privateKey = parsedSecret.privateKey;
-
-          if (userPIN && privateKey && privateKey.includes('cipher')) {
-            const decryptedPrivateKey = await encryptor.decrypt(userPIN, privateKey);
-            processedSecrets[key] = JSON.stringify({
-              ...parsedSecret,
-              privateKey: decryptedPrivateKey,
-            });
-          }
+        if ((theKeyIsASeedPhrase || theKeyIsAPrivateKey) && secret?.includes('cipher')) {
```

Comment:
The issue was that this code assumed the seed or pkey were encrypted as a stringified JSON object like `{"seedphrase": "<encrypted_data>"}`, but it is actually just the encrypted data.